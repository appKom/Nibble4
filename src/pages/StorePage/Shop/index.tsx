import React, { FC } from "react";
import styled from "styled-components";
import Items from "./Items";
import Basket from "./Basket";

const Shop: FC = () => {
  return (
    <Container>
      <Items />
      <Basket />
    </Container>
  );
};

export default Shop;

const Container = styled.div`
  display: grid;
  grid-template-columns: 2fr 1fr;
  width: 100%;
  height: 100%;
  padding: 0.5rem 0;
`;
