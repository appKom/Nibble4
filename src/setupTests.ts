// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import { CartItem, createCartItem } from "types/inventory";
import { modalTypes } from "types/modal";
import { Product } from "types/inventory";
import { User } from "types/user";

const mockProduct: Product = {
  pk: 95,
  name: "Powerking",
  description: "Energidrikk",
  price: 10,
  image: {
    sm: "/media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png",
  },
  category: {
    pk: 3,
    name: "Drikke",
  },
};

const mockUser: User = {
  pk: 1337,
  balance: 1337,
  first_name: "SuperAdmin",
};

const mockCartItem: CartItem = createCartItem(95);

const mockStateWithUser: State = {
  user: mockUser,
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
};

const mockState: State = {
  user: mockUser,
  inventory: [mockProduct],
  category: "Alt",
  cart: { 95: mockCartItem },
  modalState: modalTypes.DISABLED,
};

const emptyState: State = {
  user: null,
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
};

export const userContext: StateContextProps = {
  state: mockStateWithUser,
  dispatch: () => null,
};

export const fullContext: StateContextProps = {
  state: mockState,
  dispatch: () => null,
};

export const emptyContext: StateContextProps = {
  state: emptyState,
  dispatch: () => null,
};
