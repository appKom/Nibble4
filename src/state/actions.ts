import { Product } from "types/inventory";
import { User } from "types/user";
import { modalTypes } from "types/modal";

export enum ActionTypes {
  SET_USER = "SET_USER",
  LOGOUT_USER = "LOGOUT_USER",
  SET_INVENTORY = "SET_INVENTORY",
  SET_FAVOURITES = "SET_FAVOURITES",
  ADD_TO_CART = "ADD_TO_CART",
  REMOVE_FROM_CART = "REMOVE_FROM_CART",
  EMPTY_CART = "EMPTY_CART",
  SET_CATEGORY = "SET_CATEGORY",
  SET_MODAL_STATE = "SET_MODAL_STATE",
}

type SetUserAction = { type: ActionTypes.SET_USER; payload: User };
type LogoutUserAction = { type: ActionTypes.LOGOUT_USER };
type SetInventoryAction = {
  type: ActionTypes.SET_INVENTORY;
  payload: Product[];
};

type SetFavouritesAction = {
  type: ActionTypes.SET_FAVOURITES;
  payload: Product[];
};

type SetCategoryAction = {
  type: ActionTypes.SET_CATEGORY;
  payload: string;
};
type AddToCartAction = {
  type: ActionTypes.ADD_TO_CART;
  payload: number;
};
type RemoveFromCartAction = {
  type: ActionTypes.REMOVE_FROM_CART;
  payload: number;
};

type EmptyCartAction = {
  type: ActionTypes.EMPTY_CART;
};

type SetModalState = {
  type: ActionTypes.SET_MODAL_STATE;
  payload: modalTypes;
};

export type Action =
  | SetUserAction
  | LogoutUserAction
  | SetInventoryAction
  | SetFavouritesAction
  | SetCategoryAction
  | AddToCartAction
  | RemoveFromCartAction
  | EmptyCartAction
  | SetModalState;

export const setUser = (user: User): SetUserAction => ({
  type: ActionTypes.SET_USER,
  payload: user,
});
export const logoutUser = (): LogoutUserAction => ({
  type: ActionTypes.LOGOUT_USER,
});
export const setInventory = (inventory: Product[]): SetInventoryAction => ({
  type: ActionTypes.SET_INVENTORY,
  payload: inventory,
});

export const setFavourites = (favourites: Product[]): SetFavouritesAction => ({
  type: ActionTypes.SET_FAVOURITES,
  payload: favourites,
});

export const addToCart = (id: number): AddToCartAction => ({
  type: ActionTypes.ADD_TO_CART,
  payload: id,
});
export const removeFromCart = (id: number): RemoveFromCartAction => ({
  type: ActionTypes.REMOVE_FROM_CART,
  payload: id,
});
export const emptyCart = (): EmptyCartAction => ({
  type: ActionTypes.EMPTY_CART,
});

export const setCategory = (id: string): SetCategoryAction => ({
  type: ActionTypes.SET_CATEGORY,
  payload: id,
});

export const setModalState = (modalType: modalTypes): SetModalState => ({
  type: ActionTypes.SET_MODAL_STATE,
  payload: modalType,
});
