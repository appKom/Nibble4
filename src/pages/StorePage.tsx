import React, { FC, useContext } from "react";
import { Redirect } from "react-router-dom";
import styled from "styled-components";
import { StateContext } from "state/state";
import Navbar from "components/Shop/Navbar";
import Window from "components/Shop/Window";
import { initialRoute } from "utility/routes";

const StorePage: FC = () => {
  const { state } = useContext(StateContext);
  if (!state.user) return <Redirect to={initialRoute} />;
  return (
    <Container>
      <Navbar />
      <Window />
    </Container>
  );
};

const Container = styled.div``;

export default StorePage;
