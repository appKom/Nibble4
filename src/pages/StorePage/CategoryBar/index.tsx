import React, { useContext, FC } from "react";
import styled from "styled-components";
import Category from "./Category";
import { StateContext } from "state/state";
import { favouritesCategory, getCategories } from "types/inventory";
import { setCategory } from "state/actions";

const CategoryBar: FC = () => {
  const { state, dispatch } = useContext(StateContext);
  const categories = getCategories(state.inventory, state.user!);

  // If the logged in user has a substantial amount of products previously purchased,
  // automatically set the category to "Dine favoritter".
  if (state.user!.favourites.length >= 4)
    dispatch(setCategory(favouritesCategory));

  return (
    <Container>
      {categories.map((category) => (
        <Category key={category} category={category} />
      ))}
    </Container>
  );
};

export default CategoryBar;

const Container = styled.div`
  display: flex;
  margin-left: 25px;
  height: 30px;
`;
