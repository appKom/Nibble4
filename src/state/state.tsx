import React, { createContext, Dispatch, ReactNode, useReducer } from "react";
import {
  Product,
  CartItem,
  incrementCartItem,
  createCartItem,
  decrementCartItem,
} from "types/inventory";
import { User, OlCoinsUser } from "types/user";
import { modalTypes } from "types/modal";
import { Action, ActionTypes } from "state/actions";

type State = {
  user?: User;
  olCoinsUser?: OlCoinsUser;
  inventory: Product[];
  category: string;
  cart: { [id: number]: CartItem };
  modalState: modalTypes;
  newOlCoins: number;
};

const initialState: State = {
  olCoinsUser: { balance: 100 },
  inventory: [],
  category: "Alt",
  cart: {},
  modalState: modalTypes.DISABLED,
  newOlCoins: 0,
};

const getIncrementedCartItem = (id: number, state: State): CartItem => {
  if (state.cart[id]) {
    return incrementCartItem(state.cart[id]);
  }
  return createCartItem(id);
};

const getDecrementedCart = (id: number, state: State) => {
  if (state.cart[id].quantity > 1) {
    return { ...state.cart, [id]: decrementCartItem(state.cart[id]) };
  }
  const removedCart = { ...state.cart };
  delete removedCart[id];
  return { ...removedCart };
};

const updateNewOlCoins = (newValue: number, state: State) => {
  return state.newOlCoins + newValue;
};

const StateReducer = (state: State, action: Action): State => {
  switch (action.type) {
    case ActionTypes.SET_USER:
      return { ...state, user: action.payload };
    case ActionTypes.LOGOUT_USER:
      return {
        ...state,
        user: undefined,
        cart: {},
        category: "Alt",
        modalState: modalTypes.DISABLED,
      };
    case ActionTypes.SET_INVENTORY:
      return { ...state, inventory: action.payload };
    case ActionTypes.SET_CATEGORY:
      return { ...state, category: action.payload };
    case ActionTypes.SET_MODAL_STATE:
      return { ...state, modalState: action.payload };
    case ActionTypes.CHANGE_NEWOLCOINS:
      return {
        ...state,
        newOlCoins: updateNewOlCoins(action.payload, state),
      };
    case ActionTypes.ADD_TO_CART:
      return {
        ...state,
        cart: {
          ...state.cart,
          [action.payload]: getIncrementedCartItem(action.payload, state),
        },
      };
    case ActionTypes.REMOVE_FROM_CART: {
      return {
        ...state,
        cart: getDecrementedCart(action.payload, state),
      };
    }
    case ActionTypes.EMPTY_CART:
      return { ...state, cart: {} };
    default:
      return { ...state };
  }
};

type StateContextProps = {
  state: State;
  dispatch: Dispatch<Action>;
};

const initialContext: StateContextProps = {
  state: initialState,
  dispatch: () => null,
};

export const StateContext = createContext<StateContextProps>(initialContext);

type StateProviderProps = {
  children: ReactNode;
};

export const StateProvider = (props: StateProviderProps): JSX.Element => {
  const [state, dispatch] = useReducer(StateReducer, initialState);
  const { children } = props;

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {children}
    </StateContext.Provider>
  );
};
