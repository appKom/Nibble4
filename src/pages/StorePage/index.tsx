import React, { FC, useContext } from "react";
import { Redirect } from "react-router-dom";
import { StateContext } from "state/state";
import styled from "styled-components";
import TitleBar from "./TitleBar";
import CategoryBar from "./CategoryBar";
import Shop from "./Shop";
import { initialRoute } from "utility/routes";

const StorePage: FC = () => {
  const { state } = useContext(StateContext);
  if (!state.user) return <Redirect to={initialRoute} />;
  return (
    <Container>
      <TitleBar />
      <CategoryBar />
      <Shop />
    </Container>
  );
};

const Container = styled.div`
  height: 100%;
`;

export default StorePage;
