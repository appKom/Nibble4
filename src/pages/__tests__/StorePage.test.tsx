import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import StorePage from "../StorePage";
import { StateContext } from "state/state";
import { fullContext } from "setupTests";

afterEach(cleanup);

it("Should match snapshot", () => {
  const { asFragment } = render(
    <StateContext.Provider value={fullContext}>
      <StorePage />
    </StateContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});
