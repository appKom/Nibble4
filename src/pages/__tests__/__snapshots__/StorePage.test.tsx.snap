// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-fodVxV lbybPD"
  >
    <div
      class="sc-bdfBwQ kZrORh"
    >
      <img
        alt="Online logo"
        src="/images/logo.png"
      />
      <div>
        <span>
          SuperAdmin | 
          <svg
            fill="currentColor"
            height="20px"
            stroke="currentColor"
            stroke-width="0"
            style="color: white;"
            viewBox="0 0 512 512"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M200.4 27.39L180.9 183h42.8l49.1-146.57-72.4-9.04zm91.7 8L242.7 183l149.7.1 34.3-102.61-134.6-45.1zM180 46.03l-71.9 7.84L122.2 183h40.7L180 46.03zM64 153c-11.5 0-19.18 8.8-21.27 17.2-1.04 4.2-.45 7.6.73 9.5 1.17 1.8 2.79 3.3 8.54 3.3h52.1l-3.3-30H64zm357.4 0l-10 30h47.5c-2.6-5-3.7-10.3-3-15.6.7-5.2 2.7-9.9 5.3-14.4h-39.8zM41 201v246.9c0 5.1 2.79 11.1 7.37 15.7C52.96 468.2 59 471 64 471l384 .1c5 0 11-2.8 15.6-7.4 4.6-4.6 7.4-10.6 7.4-15.7v-71h-87c-44 0-44-82 0-82h87v-93.9L41 201zm343 112c-20 0-20 46 0 46h22.3c-9-3.8-15.3-12.7-15.3-23s6.3-19.2 15.3-23H384zm41.7 0c9 3.8 15.3 12.7 15.3 23s-6.3 19.2-15.3 23H487v-46h-61.3zm-9.7 16c-4 0-7 3-7 7s3 7 7 7 7-3 7-7-3-7-7-7z"
            />
          </svg>
           1337kr
        </span>
      </div>
      <svg
        fill="currentColor"
        height="40px"
        stroke="currentColor"
        stroke-width="0"
        style="color: white;"
        viewBox="0 0 24 24"
        width="40px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M5 22a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v3h-2V4H6v16h12v-2h2v3a1 1 0 0 1-1 1H5zm13-6v-3h-7v-2h7V8l5 4-5 4z"
          />
        </g>
      </svg>
    </div>
    <div
      class="sc-hBEYos cQbDIC"
    >
       
      <button
        class="sc-kstrdz hFwqZf"
        style="background-color: rgb(249, 183, 89);"
      >
        Alt
      </button>
      <button
        class="sc-kstrdz hFwqZf"
        style="background-color: rgb(13, 84, 116);"
      >
        Drikke
      </button>
       
    </div>
    <div
      class="sc-bqyKva eMKDgV"
    >
      <div
        class="sc-hKgILt flfMrp"
      >
        <div
          class="sc-gsTCUz irNvkl"
        >
          <img
            alt="Product image"
            src="https://online.ntnu.no//media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png"
          />
          <div
            class="sc-dlfnbm iVQmJT"
          >
            <span>
              <b>
                 Powerking 
              </b>
            </span>
            <span>
              <b>
                10
              </b>
              kr
            </span>
          </div>
        </div>
      </div>
      <div
        class="sc-iqHYGH cSvoCr"
      >
        <h2>
           Handlekurven din
        </h2>
        <div
          class="sc-crrsfI fmjLHG"
        >
          <div
            class="sc-eCssSg impzwC"
          >
            <img
              alt="Product image"
              src="https://online.ntnu.no//media/images/responsive/sm/38688f77-21b6-4936-8ba6-c976538fb110.png"
            />
            <span>
              <b>
                 Powerking 
              </b>
            </span>
            <span />
            <span>
              Quantity:
              <b>
                1
              </b>
            </span>
            <span>
              <b>
                10
              </b>
              kr
            </span>
          </div>
        </div>
        <div
          class="sc-dQppl icFKbN"
        >
          <span>
            <b>
               Sum :
            </b>
          </span>
          <span
            id="pris"
          >
            <b>
               10kr
            </b>
          </span>
        </div>
        <button
          class="sc-jSgupP chIaFM"
        >
          Kj√∏p
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
